internal/controller/dashboard_controller.go

```diff
@@ -206,23 +206,15 @@ func (r DashboardReconciler) configureDashboard(ctx context.Context, dashboardCM
 		return nil
 	}
 
-	// We always switch back to the shared org
-	defer func() {
-		if _, err = r.GrafanaAPI.SignedInUser.UserSetUsingOrg(grafana.SharedOrg.ID); err != nil {
-			logger.Error(err, "failed to change current org for signed in user")
-		}
-	}()
-
 	// Switch context to the dashboards-defined org
 	organization, err := grafana.FindOrgByName(r.GrafanaAPI, dashboardOrg)
 	if err != nil {
 		logger.Error(err, "failed to find organization", "organization", dashboardOrg)
 		return errors.WithStack(err)
 	}
-	if _, err = r.GrafanaAPI.SignedInUser.UserSetUsingOrg(organization.ID); err != nil {
-		logger.Error(err, "failed to change current org for signed in user")
-		return errors.WithStack(err)
-	}
+	currentOrgID := r.GrafanaAPI.OrgID()
+	r.GrafanaAPI.WithOrgID(organization.ID)
```

Comment:
This function does nothing to api calls at all and we need to use the exisiting functions.
Now I do see the issue with race conditions 🤔

Reply:
At least we tested this with Zirko when using the organizations Apis. Maybe it's différent with the dashboard api

Reply:
This does add an `X-Grafana-Org-Id` header to requests https://github.com/grafana/grafana-openapi-client-go/blob/8d7e1f158f65542f533d67b8fc335f4491a62a39/client/grafana_http_api_client.go#L448

I tested this and I could see datasources being attached to multiple organizations.

There's a race condition when using `SignedInUser.UserSetUsingOrg` because this does switch the organization context for the user on the server side, so multiple clients using the same credentials would be bound to the same server organization context.

Reply:
So it might work for datasources and dashboards sure but the grafana api is not Always supporting that query param

Reply:
Well not all apis need a org context. So I assume the client does the right thing.

Reply:
Would you mind splitting this change in another PR and make sure it's well tested then? If that works then i'm fine with that :)

Reply:
So, I guess:
- let's use it where we can because it's better than the server-side setting
- make sure to write comments in the code where we can't

Then when we have a clear view of what resources require server-side setting of the org, we can think about how to prevent race conditions. I suspect we need to have them all using a common and unique client + thread.

Reply:
And I agree this would be easier to review and merge the multiple changes from if they were in separate PRs.

Reply:
As long as it's set on the server side and we use the same credentials then using a différent client of thread will not really help though :)

Reply:
Merged in https://github.com/giantswarm/observability-operator/pull/252